<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <link href="docs.css" rel="stylesheet">
    <title>How to create an alignment</title>
</head>
<body>
<div id="docHeader"><a href="http://www.genewarrior.com"><img src="../css/genewarrior_mansgoggel.png"><img id="logo"
                                                                                                           src="../css/genewarrior_dnatools.png"></a>
    <br>This page is part of the GeneWarrior Documentation. Go to the <a href="../index">main site</a> of GeneWarrior
</div>
<h2>How to create an Alignment</h2>
<p>See also <a href="exp_alignment.html">What is an Alignment?</a></p>

<ol type="1">
    <li>Enter your Protein or DNA sequences into the text box and click the button "<i>Add</i>". Your new entry is
        added.<br>
        Enter at least two sequences to align together. If you don't have any sequences and just want to try it out, you
        can import some example sequences by clicking on "<i>Load Examples</i>", next to the "<i>Add</i>"-Button.<br>
        You can enter your sequences one by one or as Multi-FASTA together. See <a href="doc_sequenceimport.html">Help>Sequence
            Import</a> for more details
        on how to import sequences.
        <img src="pics/tut_createalignment_1.png"><br>
    </li>
    <li>Select all your entries that you want to convert by clicking on them. Selected entries change the background
        color.<br>
        <img src="pics/tut_createalignment_2.png"><br>
        To deselect an entry, click on it again; the background changes to white again. Select at least two sequences to
        produce an alignment.<br>
        <small>For more details about selecting entries see <a href="doc_selecting.html">Help>Selecting entries</a><br>
            <a href="doc_rename.html">Help>Renaming entries</a> describes how to change an entry name.</small>

    </li>
    <li>After selecting your entries, choose the appropriate command in the tool bar: "<i>Create Alignment > MUSCLE</i>".<br>
        <img src="pics/tut_createalignment_3.png"><br>
        <small>For a detailed explanation of the options "<i>Full sequence</i> (CFE Alignment)", "<i>Highest similarity
            region</i> (Local Alignment)" and "<i>MUSCLE</i> (Multiple Sequence Alignment)", see the <a
                href="doc_createalignment.html">Tool documentation</a>.</small><br>
        A new alignment entry is added to your project.<br>
        <img src="pics/tut_createalignment_4.png"><br>
        The yellow markings represent conservation of the respective column.
    </li>
    <li>
        Alignments can be furthered analyzed by creating consensus sequences, phylogenetic trees or a sequence logo.
        This is described on the next tutorial: <a href="tut_analyzealignment.html">Tutorial > Analyze Alignments</a>.
    </li>
</ol>

<p>For an explanation about Alignments see <a href="exp_alignment.html">What is an Alignment?</a></p>
<p id="gotoIndex"><a href="index.html">Back to Documentation index</a></p>
</body>
</html>
